clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -84,26 +96,266 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+
+    mlir::Type srcTy = src.getType();
+
+    // Handle conversions to bool first, they are special: comparisons against
+    // 0.
+    if (dstType->isBooleanType())
+      return emitConversionToBool(src, srcType, cgf.getLoc(loc));
+
+    mlir::Type dstTy = convertType(dstType);
+
+    if (srcType->isHalfType() &&
+        !cgf.getContext().getLangOpts().NativeHalfType) {
+      // Cast to FP using the intrinsic if the half type itself isn't supported.
+      if (mlir::isa<cir::CIRFPTypeInterface>(dstTy)) {
+        if (cgf.getContext().getTargetInfo().useFP16ConversionIntrinsics())
+          cgf.getCIRGenModule().errorNYI(loc,
+                                         "cast via llvm.convert.from.fp16");
+      } else {
+        // Cast to other types through float, using either the intrinsic or
+        // FPExt, depending on whether the half type itself is supported (as
+        // opposed to operations on half, available with NativeHalfType).
+        if (cgf.getContext().getTargetInfo().useFP16ConversionIntrinsics()) {
+          cgf.getCIRGenModule().errorNYI(loc,
+                                         "cast via llvm.convert.from.fp16");
+        } else {
+          src = builder.createCast(cgf.getLoc(loc), cir::CastKind::floating,
+                                   src, cgf.FloatTy);
+        }
+        srcType = cgf.getContext().FloatTy;
+        srcTy = cgf.FloatTy;
+      }
+    }
+
+    // TODO(cir): LLVM codegen ignore conversions like int -> uint,
+    // is there anything to be done for CIR here?
+    if (srcTy == dstTy) {
+      if (opts.emitImplicitIntegerSignChangeChecks)
+        cgf.getCIRGenModule().errorNYI(loc,
+                                       "implicit integer sign change checks");
+      return src;
+    }
+
+    // Handle pointer conversions next: pointers can only be converted to/from
+    // other pointers and integers. Check for pointer types in terms of LLVM, as
+    // some native types (like Obj-C id) may map to a pointer type.
+    if (auto dstPT = dyn_cast<cir::PointerType>(dstTy)) {
+      cgf.getCIRGenModule().errorNYI(loc, "pointer casts");
+    }
+
+    if (isa<cir::PointerType>(srcTy)) {
+      // Must be an ptr to int cast.
+      assert(isa<cir::IntType>(dstTy) && "not ptr->int?");
+      return builder.createPtrToInt(src, dstTy);
+    }
+
+    // A scalar can be splatted to an extended vector of the same element type
+    if (dstType->isExtVectorType() && !srcType->isVectorType()) {
+      // Sema should add casts to make sure that the source expression's type
+      // is the same as the vector's element type (sans qualifiers)
+      assert(dstType->castAs<ExtVectorType>()->getElementType().getTypePtr() ==
+                 srcType.getTypePtr() &&
+             "Splatted expr doesn't match with vector element type?");
+
+      llvm_unreachable("not implemented");
+    }
+
+    if (srcType->isMatrixType() && dstType->isMatrixType())
+      cgf.getCIRGenModule().errorNYI(loc,
```

Comment:
here too.  it ends up just falling through to the assert below which is awkward.

Reply:
still valid here, I think we need to just generate a constant of the dest type.

Reply:
perhaps a `return nullptr' here?